### PR TITLE
fix: fix eval in multi interpreter borrow

### PR DESCRIPTION
**Pull request by dranikpg**

### CODE REVIEW

File: src/server/main_service.cc [hunk 1]
```diff
@@ -1821,6 +1840,9 @@ void Service::Exec(CmdArgList args, ConnectionContext* cntx) {
   SinkReplyBuilder::ReplyAggregator agg(rb);
   rb->StartArray(exec_info.body.size());
 
+  if (state != ExecEvalState::NONE)
```
Comment (romange) [inline src/server/main_service.cc:1843]: why not test for a specific state?

File: src/server/main_service.cc [hunk 2]
```diff
@@ -605,6 +605,31 @@ optional<Transaction::MultiMode> DeduceExecMode(ExecEvalState state,
   return multi_mode;
 }
 
+// Either take the interpreter from the preborrowed multi exec transaction or borrow one.
+struct BorrowedInterpreter {
```
Comment (romange) [inline src/server/main_service.cc:609]: nice! maybe convert to a class and hide data members?

File: src/server/main_service.cc [hunk 3]
```diff
@@ -605,6 +605,31 @@ optional<Transaction::MultiMode> DeduceExecMode(ExecEvalState state,
...
+  BorrowedInterpreter(ConnectionContext* cntx) {
```
Comment (romange) [inline src/server/main_service.cc:610]: explicit

File: src/server/main_service.cc (hunk 1 above)
Comment (romange) [reply src/server/main_service.cc:1843]: do we borrow for every multi-exec transaction?

Comment (dranikpg) [reply src/server/main_service.cc:1843]: `ExecEvalState` show whether it contains any eval, we have ALL, SOME, and NONE. So if we have at least one (i.e. not NONE) we borrow

File: src/server/main_service.cc [hunk 4]
```diff
@@ -605,6 +605,32 @@ optional<Transaction::MultiMode> DeduceExecMode(ExecEvalState state,
...
+    if (owned)
+      ServerState::tlocal()->ReturnInterpreter(interpreter);
+  }
+
+  operator Interpreter*() {
+    return interpreter;
+  }
+
+ private:
+  Interpreter* interpreter = nullptr;
```
Comment (romange) [inline src/server/main_service.cc:630]: underscores :)

**Review by romange — APPROVED** [review]

**Pull request merged**